docs/log-correlation.asciidoc

```diff
@@ -0,0 +1,119 @@
+[[log-correlation]]
+== Log Correlation
+
+The Go agent provides integrations for popular logging frameworks, which take care of
+injecting trace ID fields into your application's log records. You can find a list of
+the supported logging frameworks under <<supported-tech-logging, supported technologies>>.
+
+If your favorite logging framework is not already supported, there are two other options:
+
+* Open a feature request, or contribute code, for additional support, as described in <<contributing>>.
+* Manually inject trace IDs into log records, as described below in <<log-correlation-manual>>.
+
+Regardless of how you integrate APM with logging, you can use {filebeat-ref}[Filebeat] to
+send your logs to Elasticsearch, in order to correlate your traces and logs and link from
+APM to the {kibana-ref}/xpack-logs.html[Logs app].
+
+[float]
+[[log-correlation-manual]]
+=== Manual log correlation
+
+If the agent-provided logging integrations are not suitable or not available for your
+application, then you can use the agent's <<api, API>> to inject trace IDs manually.
+There are two main approaches you can take, depending on whether you are using structured
+or unstructured logging.
+
+[float]
+[[log-correlation-manual-structured]]
+==== Manual log correlation (structured)
+
+For correlating structured logs with traces, the following fields should be added to your logs:
+
+ - trace.id
+ - transaction.id
+ - span.id (if logging in the context of a span)
+
+Given a transaction object, you can obtain its trace ID and transaction ID using
+the <<transaction-tracecontext, apm.Transaction.TraceContext>> method. Similarly,
+given a span object, you can obtain its span ID using <<span-tracecontext, apm.Span.TraceContext>>.
+
+If you use the context APIs to start transactions and spans, then you can obtain
+the context's current transaction using <<apm-transaction-from-context, apm.TransactionFromContext>>,
+and current span using <<apm-span-from-context, apm.SpanFromContext>>. Note that if
+a transaction is not sampled, `apm.TransactionFromContext` will return `nil`.
+Similarly, spans may be dropped by the agent, so `apm.SpanFromContext` may also return nil.
+
+[source,go]
+----
+labels := make(map[string]string)
+tx := apm.TransactionFromContext(ctx)
+if tx != nil {
+	traceContext := tx.TraceContext()
+	labels["trace.id"] = traceContext.Trace.String()
+	labels["transaction.id"] = traceContext.Span.String()
+	if span := apm.SpanFromContext(ctx); span != nil {
+		labels["span.id"] = span.TraceContext().Span
+	}
+}
+----
+
+You can follow this article in order to ingest JSON-encoded logs with Filebeat:
+{blog-ref}how-to-instrument-your-go-app-with-the-elastic-apm-go-agent#logs[How to instrument your Go app with the Elastic APM Go agent].
```

Comment:
I agree that the content of this link is better suited for a blog post than documentation, but I'm wondering if you think it's possible/useful to include some of this information in the docs?

Reply:
I would love to see more guided walkthroughs in our docs in general. For the specific case of how to ingest JSON logs using Filebeat, to me it seems more appropriate to add it to FIlebeat docs than here. WDYT?

Reply:
Yeah, it's a tough line to toe. Our docs are so difficult to navigate cross-stack that I worry about users getting lost along the way. With this in mind, I've added a small amount of Filebeat information to the central documentation for this feature -- with the hopes that it will be helpful enough for users, and not be too difficult to maintain.

https://github.com/elastic/apm-server/pull/2672

Reply:
Cool, I'll link to that instead then.

Reply:
... of course that's not actually merged yet, so I can't. I'll update the Go docs once that's merged and released.